src/tribler-core/tribler_core/modules/popularity/popularity_community.py

```diff
@@ -49,23 +50,28 @@ def gossip_torrents_health(self):
         self.ez_send(random_peer, TorrentsHealthPayload.create(random_torrents_checked, popular_torrents_checked))
 
     @lazy_wrapper(TorrentsHealthPayload)
-    async def on_torrents_health(self, _, payload):
+    async def on_torrents_health(self, peer, payload):
         self.logger.info("Received torrent health information for %d random torrents and %d checked torrents",
                          len(payload.random_torrents), len(payload.torrents_checked))
         all_torrents = payload.random_torrents + payload.torrents_checked
 
-        def _put_health_entries_in_db():
-            with db_session:
-                for infohash, seeders, leechers, last_check in all_torrents:
-                    torrent_state = self.metadata_store.TorrentState.get(infohash=infohash)
-                    if torrent_state and last_check > torrent_state.last_check:
-                        # Replace current information
-                        torrent_state.seeders = seeders
-                        torrent_state.leechers = leechers
-                        torrent_state.last_check = last_check
-                    elif not torrent_state:
-                        _ = self.metadata_store.TorrentState(infohash=infohash, seeders=seeders,
-                                                             leechers=leechers, last_check=last_check)
-
-            self.metadata_store.disconnect_thread()
-        await get_event_loop().run_in_executor(None, _put_health_entries_in_db)
+        await self.put_health_entries_in_db(peer, all_torrents)
+
+    @db_session
```

Comment:
Does this work with async function? @kozlovsky

Reply:
No, it should not work. It is necessary to put db_session inside:

```python
    async def put_health_entries_in_db(self, peer, all_torrents):
        with db_session:
            for infohash, seeders, leechers, last_check in all_torrents:
```
(and never call functions with `await` inside db_session context manager body)

Reply:
@xoriole @kozlovsky 
Fixed. Thanks for pointing this out!

Please check the current implementation:
```
 @lazy_wrapper(TorrentsHealthPayload)
    async def on_torrents_health(self, peer, payload):
        self.logger.info("Received torrent health information for %d random torrents and %d checked torrents",
                         len(payload.random_torrents), len(payload.torrents_checked))
        all_torrents = payload.random_torrents + payload.torrents_checked

        with db_session:
            for infohash, seeders, leechers, last_check in all_torrents:
                torrent_state = self.metadata_store.TorrentState.get(infohash=infohash)
```

Reply:
This should work. The idea is that `db_session` should never span asynchronous functions.